File: backend/src/models/storagezone.py
from pydantic import BaseModel, Field, ConfigDict, model_validator
from typing import Optional
from datetime import datetime
from src.database.db import ZoneStatus


class StorageZoneBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    grain_type_id: int = Field(..., gt=0)
    total_capacity: int = Field(..., gt=0, description="Total capacity in kg (kilograms)")
    available_capacity: int = Field(..., ge=0, description="Available capacity in kg (kilograms)")
    status: ZoneStatus

    @model_validator(mode="after")
    def validate_capacity(self):
        if self.available_capacity > self.total_capacity:
            raise ValueError("available_capacity cannot exceed total_capacity")
        return self


class StorageZoneCreate(StorageZoneBase):
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "name": "Zone A - Wheat Storage",
                "grain_type_id": 1,
                "total_capacity": 10000,  # in kg
                "available_capacity": 5000,  # in kg
                "status": "active",
            }
        }
    )


class StorageZoneUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    grain_type_id: Optional[int] = Field(None, gt=0)
    total_capacity: Optional[int] = Field(None, gt=0, description="Total capacity in kg (kilograms)")
    available_capacity: Optional[int] = Field(None, ge=0, description="Available capacity in kg (kilograms)")
    status: Optional[ZoneStatus] = None
    warehouse_id: Optional[int] = Field(None, gt=0)

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "name": "Updated Zone Name",
                "available_capacity": 3000,  # in kg
                "status": "not_active",
            }
        }
    )


class StorageZoneResponse(StorageZoneBase):
    zone_id: int
    warehouse_id: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: mobile/lib/models/storage_zone.dart
class StorageZone {
  final int zoneId;
  final int warehouseId;
  final int grainTypeId;
  final String name;
  final int totalCapacity;
  final int availableCapacity;
  final String status;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  StorageZone({
    required this.zoneId,
    required this.warehouseId,
    required this.grainTypeId,
    required this.name,
    required this.totalCapacity,
    required this.availableCapacity,
    required this.status,
    this.createdAt,
    this.updatedAt,
  });

  factory StorageZone.fromJson(Map<String, dynamic> json) {
    // Helper function to safely convert to int
    int _toInt(dynamic value) {
      if (value == null) return 0;
      if (value is int) return value;
      if (value is String) return int.tryParse(value) ?? 0;
      if (value is double) return value.toInt();
      return 0;
    }

    return StorageZone(
      zoneId: _toInt(json['zone_id'] ?? json['zoneId']),
      warehouseId: _toInt(json['warehouse_id'] ?? json['warehouseId']),
      grainTypeId: _toInt(json['grain_type_id'] ?? json['grainTypeId']),
      name: json['name']?.toString() ?? '',
      totalCapacity: _toInt(json['total_capacity'] ?? json['totalCapacity'] ?? 0),
      availableCapacity: _toInt(json['available_capacity'] ?? json['availableCapacity'] ?? 0),
      status: json['status']?.toString() ?? 'active',
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'].toString())
          : null,
      updatedAt: json['updated_at'] != null
          ? DateTime.tryParse(json['updated_at'].toString())
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'zone_id': zoneId,
      'warehouse_id': warehouseId,
      'grain_type_id': grainTypeId,
      'name': name,
      'total_capacity': totalCapacity,
      'available_capacity': availableCapacity,
      'status': status,
      'created_at': createdAt?.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
    };
  }

  double get capacityPercentage => totalCapacity > 0 
      ? (availableCapacity / totalCapacity) * 100 
      : 0.0;
}

